pkg/queue/breaker.go

```diff
@@ -60,12 +65,84 @@ func (b *Breaker) Maybe(thunk func()) bool {
 	case b.pendingRequests <- t:
 		// Pending request has capacity.
 		// Wait for capacity in the active queue.
-		b.activeRequests <- t
+		b.sem.Get()
 		// Defer releasing capacity in the active and pending request queue.
-		defer func() { <-b.activeRequests; <-b.pendingRequests }()
+		defer func() { b.sem.Put(); <-b.pendingRequests }()
 		// Do the thing.
 		thunk()
 		// Report success
 		return true
 	}
 }
+
+// NewSemaphore creates a semaphore with the desired maximal and initial capacity
+func NewSemaphore(maxCapacity, initialCapacity int32) *Semaphore {
+	if initialCapacity < 0 || initialCapacity > maxCapacity {
+		panic(fmt.Sprintf("Initial capacity must be between 0 and maximal capacity. Got %v.", initialCapacity))
+	}
+	queue := make(chan token, maxCapacity)
+	sem := Semaphore{queue: queue}
+	if initialCapacity > 0 {
+		sem.AddCapacity(initialCapacity)
+	}
+	return &sem
+}
+
+// Semaphore is an implementation of a semaphore based on Go channels
+// The number of available tokens is the number of elements in the buffered channel
+type Semaphore struct {
+	queue    chan token
+	token    token
+	reducers int32
+	capacity int32
```

Comment:
What's capacity needed for? Seems like it's not used anywhere but to increment and decrement it?

Reply:
I use it in one of the tests and also as a guard to prevent reducing more capacity than we added (just added)